### ISSUE TITLE
Add loading animation to splash

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/joist#68]
Similar to the 'infinte scrolling' progress bar?

**Comment by samreid** [comment]
We noticed that the simulations at http://connexions.github.io/simulations/ have a nice animated splash screen.  We should look into how they did it and if we can do something similar.

**Comment by samreid** [comment]
The connexions splash has unreliable behavior on the iPad3.  Sometimes it animates, sometimes it is blank, sometimes the beaker appears with no bubbles, etc.

@jakies pointed out the code that renders it is in CSS and located here:
https://github.com/Connexions/simulations/blob/master/common/app/app.less

**Comment by ariel-phet** [comment]
Concord consortium also seems to have a loading animation on their HTML5 sims. See for instance:

http://mw.concord.org/nextgen/#interactives/chemistry/phase-change/molecular-view-liquid

**Comment by samreid** [comment]
I also saw this loading spinner, which seemed immune to blocking JS in the console (at least in Chrome) https://github.com/fgnass/spin.js

**Comment by ariel-phet** [comment]
A suggestion floated to me (which I am curious about) is that we could do asynchronous loading and actually ship 2 files when you click the "play" simulation button.

My housemate brought up an interesting point -- the single html file for download does not really need a loading screen (in fact you do not really even see it even on a sim like function builder) since once you have it downloaded you are good to go, it opens seamlessly without any appreciable delay (much like sims in the iPad App since they are stored locally). He claimed that with our single HTML file solution getting animation to work while the JS loads is pretty much impossible with the way things currently work. 

However, he said we might be able to have two different methods. The file for download being the single HTML file, and the file for running online containing two files, one being the loading animation. This would of course mean a change to our build process, not sure if it is practical/feasible/etc, but I am interested in investigating a bit. Since all loading screens are the same, it seems perhaps we could "tack on" this file to any HTML5 sim being run on the site. Curious to discuss.

**Comment by samreid** [comment]
> My housemate brought up an interesting point -- the single html file for download does not really need a loading screen (in fact you do not really even see it even on a sim like function builder) since once you have it downloaded you are good to go, it opens seamlessly without any appreciable delay (much like sims in the iPad App since they are stored locally).

On iPad3, it takes about 11,000 milliseconds for Function Builder to launch **after** the full HTML file has been downloaded.  On Firefox on my MacBook Air, it is taking about 5,000 ms after the full HTML file has been downloaded.

Showing a loading animation during code downloading would be helpful for situations that take a long time to download the HTML, which may include: people using a poor LTE or 3G connection, users that are far from our server (such as Europe and Asia), classrooms with poor wifi, etc.  But for all of those cases, even after the simulation is downloaded, while the initialization JS code is running, no animation can be shown.  For instance, let's say I am in Japan on a flaky LTE connection on my iPad3, then it could take 32 seconds to download function builder, then an additional 11 seconds to run the initialization sequence.

The only possible solution I am aware of for showing incremental loading during the initialization sequence is to re-architect our initialization sequence to initialize one screen at a time, and show a progress bar moving forward after each screen.  (Say, for a 4 screen simulation, it would initialize in 4 segments).  This wouldn't help sims that have only one screen, but maybe they don't take as long to initialize and hence don't need a solution here.

I'm not opposed to additionally discussing and showing some progress animation during downloading, but I thought some browsers already had something like that (for instance, the blue line that moves from left to right on iPad).

**Comment by ariel-phet** [comment]
@jbphet and @schmitzware are curious to explore this.

Current thought, a animation while HTML is downloading and progress bar as joist initiates screens (perhaps granularity of model and view)

**Comment by samreid** [comment]
Instead of a 2-file solution, I'm wondering if we can put code in the `<head>` or somewhere so it will run ASAP while the file is still downloading.  Then maybe we can show an animation while the file is still loading but still only deliver a single file (and hence minimize changes to our build/deploy/delivery process).

**Comment by samreid** [comment]
According to http://stackoverflow.com/questions/8249013/why-put-javascript-in-head, 

> Anything in the head must be completed before the body is loaded, so it is generally a bad idea to put javascript in there. If you need something while the body is loading, or want to expedite some ajax, then it would be appropriate to put it in the head.

So this may be a suitable place for us to put loading code.

**Comment by samreid** [comment]
I was curious about the idea of doing this with a single script, but putting the loading script near the top of the file.  So I added this code as the first script in the body:

```html
<script>
  var span = document.createElement('span');
  span.style.color = 'white';
  span.innerHTML = 'Loading';
  document.body.appendChild(span);
  setInterval(function(){
    span.innerHTML = span.innerHTML+'.';
  },100);
</script>
```

and throttled my connection to simulate 4G, and it looked like this:

![loadingscreen](https://cloud.githubusercontent.com/assets/679486/16666750/43675006-4446-11e6-8bfb-f92da021e713.gif)

You can see the Loading........ pause when the code is fully downloaded and the JS starts running.  Putting this animation in the same HTML (instead of a separate file) would be immensely simpler as well as for making for a faster download, so I recommend to proceed with the single file.  Of course, this approach will have to be tested on different platforms, but it seems promising.

**Comment by samreid** [comment]
I should also mention, that for this part of the animation, the sky is the limit for what we display.  It could be the airplane animating around, for example.

**Comment by samreid** [comment]
> @jbphet and @schmitzware are curious to explore this.

I'm also interested in investigating this.

**Comment by samreid** [comment]
To understand the issue for showing progress bars while the simulation is loading, I rewrote this code:

```js
    // Instantiate the screens. Currently this is done eagerly, but this pattern leaves open the door for loading things
    // in the background.
    screens.forEach( function initializeScreen( screen ) {
      screen.backgroundColorProperty.link( sim.updateBackground );
      screen.initializeModelAndView();
    } );
```

as

```js
    var span = document.createElement( 'span' );
    span.style.color = 'white';
    span.innerHTML = 'Initializing Screens... (0/'+screens.length+')';
    document.body.appendChild( span );

    var setProgressBar = function( distance, max ) {
      span.innerHTML = 'Initializing Screens... (' + distance + '/' + max + ')';
    };

    var workItems = [];
    screens.forEach( function( screen, screenIndex ) {
      workItems.push( function() {
        screen.backgroundColorProperty.link( sim.updateBackground );
        screen.initializeModelAndView();
        setProgressBar( screenIndex + 1, screens.length );
      } );
    } );
    workItems.push( function() {
      sim.finishInit( screens, options );
      sim.start();
    } );

    var runItem = function( i ) {
      setTimeout( function() {
        workItems[ i ]();
        if ( i + 1 < workItems.length ) {
          runItem( i + 1 );
        }
      }, 0 );
    };
    runItem( 0 );
```

To accomplish this, I also had to move out the last half of the constructor to a new function like so:

```js
    finishInit: function( screens, options ) {
      var sim = this;

      // ModuleIndex should always be defined.  On startup screenIndex=0 to highlight the 1st screen.
      // When moving from a screen to the homescreen, the previous screen should be highlighted

      if ( this.homeScreen ) {
        this.rootNode.addChild( this.homeScreen.view );
      }
      _.each( screens, function( screen ) {
        screen.view.layerSplit = true;
        sim.rootNode.addChild( screen.view );
      } );
      this.rootNode.addChild( this.navigationBar );

      if ( this.homeScreen ) {

        // Once both the navbar and homescreen have been added, link the PhET button positions together.
        // See https://github.com/phetsims/joist/issues/304.
        PhetButton.linkPhetButtonTransform( this.homeScreen, this.navigationBar, this.rootNode );
      }

      this.multilink( [ 'screenIndex', 'showHomeScreen' ], function( screenIndex, showHomeScreen ) {
        if ( sim.homeScreen ) {
          sim.homeScreen.view.setVisible( showHomeScreen );
        }
        for ( var i = 0; i < screens.length; i++ ) {
          screens[ i ].view.setVisible( !showHomeScreen && screenIndex === i );
        }
        sim.navigationBar.setVisible( !showHomeScreen );
        sim.updateBackground();
      } );

      // layer for popups, dialogs, and their backgrounds and barriers
      this.topLayer = new Node();
      this.rootNode.addChild( this.topLayer );

      // @private list of nodes that are "modal" and hence block input with the barrierRectangle.  Used by modal dialogs
      // and the PhetMenu
      this.modalNodeStack = new ObservableArray(); // {Node} with node.hide()

      // @public (joist-internal) Semi-transparent black barrier used to block input events when a dialog (or other popup)
      // is present, and fade out the background.
      this.barrierRectangle = new Rectangle( 0, 0, 1, 1, 0, 0, {
        fill: 'rgba(0,0,0,0.3)',
        pickable: true
      } );
      this.topLayer.addChild( this.barrierRectangle );
      this.modalNodeStack.lengthProperty.link( function( numBarriers ) {
        sim.barrierRectangle.visible = numBarriers > 0;
      } );
      this.barrierRectangle.addInputListener( new ButtonListener( {
        fire: function( event ) {
          sim.barrierRectangle.trigger0( 'startedCallbacksForFired' );
          assert && assert( sim.modalNodeStack.length > 0 );
          sim.modalNodeStack.get( sim.modalNodeStack.length - 1 ).hide();
          sim.barrierRectangle.trigger0( 'endedCallbacksForFired' );
        }
      } ) );
      options.tandem && options.tandem.createTandem( 'sim.barrierRectangle' ).addInstance( this.barrierRectangle );

      // Fit to the window and render the initial scene
      $( window ).resize( function() { sim.resizeToWindow(); } );
      this.resizeToWindow();

      // Kick off checking for updates, if that is enabled
      UpdateCheck.check();

      // @public (joist-internal) - Keep track of the previous time for computing dt, and initially signify that time
      // hasn't been recorded yet.
      this.lastTime = -1;

      // @public (joist-internal) - Bind the animation loop so it can be called from requestAnimationFrame with the right
      // this
      this.boundRunAnimationLoop = this.runAnimationLoop.bind( this );
      this.trigger0( 'simulationStarted' );

      // Signify the end of simulation startup.  Used by PhET-iO.
      this.trigger0( 'endedSimConstructor' );
    }
```

This also meant that sims mains cannot call sim.start() immediately after the constructor finishes, hence I created a "initialized" callback that can be used like this (example from Function Builder):

```js
  var options = {
    credits: {
      leadDesign: 'Amanda McGarry',
      softwareDevelopment: 'Chris Malley (PixelZoom, Inc.)',
      team: 'Amy Hanson, Karina K. R. Hensberry, Ariel Paul, Kathy Perkins,\nSam Reid, Beth Stade, David Webb',
      qualityAssurance: 'Steele Dalton, Amanda Davis, Andrea Lin, Ben Roberts'
    },
    initialized:function(sim){sim.start();}
  };
```

Once these changes were complete, it was possible to update the DOM between each screen initialization:

![loading](https://cloud.githubusercontent.com/assets/679486/16827110/b04cc8f0-4940-11e6-88f4-26fdb4de220a.gif)

This also suggests that there is a lot of work being done before the first screen initialization starts, probably by SimLauncher initializing downloaded assets, so this could be displayed as a separate "segment" in the loading bar.

**Comment by samreid** [comment]
I'd like to help out on this issue when I have time.

**Comment by samreid** [comment]
I'm going to split this into 2 separate issues: showing animation while files are downloading and showing js execution progress.

**Comment by samreid** [comment]
Let's proceed in the two issues mentioned above.
